service/history/query.go

```diff
@@ -0,0 +1,120 @@
+// The MIT License (MIT)
+//
+// Copyright (c) 2019 Uber Technologies, Inc.
+//
+// Permission is hereby granted, free of charge, to any person obtaining a copy
+// of this software and associated documentation files (the "Software"), to deal
+// in the Software without restriction, including without limitation the rights
+// to use, copy, modify, merge, publish, distribute, sublicense, and/or sell
+// copies of the Software, and to permit persons to whom the Software is
+// furnished to do so, subject to the following conditions:
+//
+// The above copyright notice and this permission notice shall be included in all
+// copies or substantial portions of the Software.
+//
+// THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
+// IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
+// FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
+// AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
+// LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
+// OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN THE
+// SOFTWARE.
+
+package history
+
+import (
+	"errors"
+	"sync"
+
+	"github.com/pborman/uuid"
+	"github.com/uber/cadence/.gen/go/shared"
+)
+
+const (
+	queryStateBuffered queryState = iota
+	queryStateCompleted
+)
+
+type (
+	queryState int
+
+	query interface {
+		getQuerySnapshot() *querySnapshot
+		getQueryTermCh() <-chan struct{}
+		completeQuery(*shared.WorkflowQueryResult) error
+	}
+
+	queryImpl struct {
+		sync.RWMutex
+
+		id          string
+		queryInput  *shared.WorkflowQuery
+		queryResult *shared.WorkflowQueryResult
+		termCh      chan struct{}
+		state       queryState
+	}
+
+	querySnapshot struct {
+		id          string
+		queryInput  *shared.WorkflowQuery
+		queryResult *shared.WorkflowQueryResult
+		state       queryState
+	}
+)
+
+func newQuery(queryInput *shared.WorkflowQuery) query {
+	return &queryImpl{
+		id:          uuid.New(),
+		queryInput:  queryInput,
+		queryResult: nil,
+		termCh:      make(chan struct{}),
+		state:       queryStateBuffered,
+	}
+}
+
+func (q *queryImpl) getQuerySnapshot() *querySnapshot {
+	q.RLock()
+	defer q.RUnlock()
+
+	return &querySnapshot{
+		id:          q.id,
+		queryInput:  q.queryInput,
+		queryResult: q.queryResult,
```

Comment:
this is returning a pointer out, i.e. caller can manipulate the query input / query result while NOT holding a lock
if the intention of this API is truly creating a snapshot, then a deep copy should be made

Reply:
This API cannot afford the overhead of making a truly deep copy because query inputs/results can be big. Here I return a shallow copy for some protection but its not perfect. Since queryRegistry and query are just an implementation detail of mutable state I think its okay to not have perfect encapsulation here.

Reply:
an suggestion is to make this function take in an action function, which will do the logic for the caller.

Reply:
snapshot should not return pointer, which will let the caller has the ability to change the internal state.

Reply:
Spoke offline we renamed this to getInternalState.